linux-gpib-user/doc/man/man5/gpib.conf.5
'\" t
.\"     Title: gpib.conf
.\"    Author: Frank Mori Hess
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 10/04/2025
.\"    Manual: 	Configuration
.\"    Source: linux-gpib 4.3.7
.\"  Language: English
.\"
.TH "GPIB\&.CONF" "5" "10/04/2025" "linux-gpib 4.3.7" "Configuration"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gpib.conf \- GPIB library configuration file
.SH "DESCRIPTION"
.PP
The library, and the administration tool
gpib_config
read their configuration information from the file gpib\&.conf\&. By default, the file is located in the sysconfdir directory configured when linux\-gpib was compiled\&. The sysconfdir is typically set to /etc or /usr/local/etc\&. A template gpib\&.conf file can be found in the util/templates/ subdirectory of the linux\-gpib package\&.
.PP
The default config file may be overridden by setting the IB_CONFIG environment variable to a custom file path\&.
.PP
The configuration file must contain one \*(Aqinterface\*(Aq entry for each of the board minors that are going to be used unless all the \*(Aqrequired\*(Aq options are specified in the command\-line invocation of the administration tool gpib_config to configure that minor\&. The file must contain an \*(Aqinterface\*(Aq entry with the \*(Aqname\*(Aq option set if you wish to open the board desciptor with
ibfind()
instead of directly using the minor as the board descriptor\&. If there is no entry for the minor in the configuration file, the required command\-line options are \*(Aqboard_type\*(Aq and \*(Aqpad\*(Aq\&. Additionally the \*(Aq\-\-system\-controller\*(Aq option must be scpecified if the board is to be system\-controller, which is the normal case\&. For example:
.sp
.if n \{\
.RS 4
.\}
.nf
  gpib_config \-\-minor 8 \-\-board\-type ni_usb_b \-\-pad 0 \-\-system\-controller
.fi
.if n \{\
.RE
.\}
.sp
The configuration file can contain zero or more \*(Aqdevice\*(Aq entries\&. \*(Aqdevice\*(Aq entries are only required if you wish to open device descriptors with
ibfind()
instead of using
ibdev()\&. Each \*(Aqdevice\*(Aq entry must have at least the \*(Aqminor\*(Aq, \*(Aqpad\*(Aq and \*(Aqname\*(Aq options specified\&. The \*(Aqminor\*(Aq of a \*(Aqdevice\*(Aq entry must have a corresponding \*(Aqinterface\*(Aq entry\&.
.PP
Several example entries, and a table summarizing the possible options follow\&.
.PP
.if n \{\
.RS 4
.\}
.nf
  /*  This comment style is permitted */

  // This is an alternate comment style
  
interface {
	minor = 0
	board_type = "ni_pci"
	pad = 0
	master = yes
}

interface {
	minor = 1
	board_type = "ines_pci"
	name = "joe"
	pad = 5
	sad = 0
	timeout = T10s
	pci_bus = 0
	pci_slot = 0xd
	master = no
}

interface {
	minor = 2
	board_type = "pcII"
	pad = 3
	sad = 0x62
	eos = 0x0d
	set\-reos = yes
	set\-bin = no
	set\-xeos = no
	set\-eot = yes
	base = 0x300
	irq  = 5
	dma  = 0
	master = no
}

device {
	minor = 0
	name = "counter"  // name that can be used with ibfind
	pad = 24
}

device {
	minor = 0
	name = "voltmeter"
	pad = 7
	sad = 110
	eos = 0xa
	set\-reos = yes
	set\-bin = no
	set\-xeos = yes
	set\-eot = no
	timeout = T1s
}
.fi
.if n \{\
.RE
.\}
.PP
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.B Table\ \&1.\ \&configuration options
.TS
allbox expand tab(:);
lB lB lB lB.
T{
option name
T}:T{
description
T}:T{
used by interface or device entries
T}:T{
required or optional
T}
.T&
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l
l l l l.
T{
base
T}:T{
Specifies the base ioport or io memory address for a
	board that lacks plug\-and\-play capability\&.
T}:T{
interface
T}:T{
optional
T}
T{
board_type
T}:T{
Specifies the type of interface board\&.  See the last
        two columns of Table
        2\&. for a list of possible board types, and the kernel
        driver module that supports them\&. There is no default\&.
T}:T{
interface
T}:T{
required
T}
T{
dma
T}:T{
Specifies the dma channel for a board that lacks plug\-and\-play
	capability\&.
T}:T{
interface
T}:T{
optional
T}
T{
eos
T}:T{
Sets the end\-of\-string byte for board or device
	descriptors obtained with ibfind()\&. By default eos = 0\&.
	See also the set\-reos, set\-bin, and set\-xeos options\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
irq
T}:T{
Specifies the interrupt level for a board that lacks
	plug\-and\-play capability\&.
T}:T{
interface
T}:T{
optional
T}
T{
master
T}:T{
Set to \*(Aqyes\*(Aq if you want the interface board to be the
	system controller of the bus\&.  There can only be one system
	controller on a bus\&. By default master = no\&.
T}:T{
interface
T}:T{
required
T}
T{
minor
T}:T{
\*(Aqminor\*(Aq specifies the minor number of the device file
	this interface board will use\&.  A \*(Aqminor\*(Aq of 0 corresponds to
	/dev/gpib0, 1 is /dev/gpib1, etc\&.  The minor number is also
	equal to the \*(Aqboard index\*(Aq which can be used as a board
	descriptor, and is passed as the first argument
	of  ibdev()
T}:T{
interface
T}:T{
required
T}
T{
name
T}:T{
The \*(Aqname\*(Aq specifies the name which can be used with
	 ibfind() to get a descriptor for the board or device
	 associated with this entry\&.
T}:T{
interface or device
T}:T{
interface optional
	  device required
T}
T{
pad
T}:T{
Specifies the primary GPIB address (valid addresses are
	0 to 30)\&.  For interfaces, this is the primary address that
	the board will be assigned when it is first brought online\&.
	For devices, this is address that will be used by device
	descriptors obtained with ibfind()\&. No default\&.
T}:T{
interface or device
T}:T{
required
T}
T{
pci_bus
T}:T{
Useful for distinguishing between multiple PCI cards\&.  If
	you have more than one PCI card that with the same \*(Aqboard_type\*(Aq, you can
	use the \*(Aqpci_bus\*(Aq and \*(Aqpci_slot\*(Aq options to specify the particular
	card you are interested in\&.
T}:T{
interface
T}:T{
optional
T}
T{
pci_slot
T}:T{
Can be used in conjunction with \*(Aqpci_bus\*(Aq to specify
	a particular pci card\&.
T}:T{
interface
T}:T{
optional
T}
T{
sad
T}:T{
Specifies the secondary GPIB address\&.  Valid values are
	0, or 0x60 to 0x7f hexadecimal (96 to 127 decimal)\&.  A value
	of 0 means secondary addressing is disabled (the default)\&.
	Secondary addresses from 0 to 31 are specified by the
	library\*(Aqs convention of adding an offset of 0x60\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
set\-bin
T}:T{
Enables 8\-bit comparisons when matching the
	end\-of\-string byte, instead of only comparing the
	7 least significant bits\&.  Only affects descriptors
	returned by ibfind(), and has same effect as setting
	the BIN bit in a
	ibeos()
	call\&. Default is set\-bin = 0\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
set\-eot
T}:T{
Enables assertion of the EOI line at the end of writes,
	for descriptors returned by ibfind()\&.  See
	ibeot()\&.
	Default set\-eoi = yes\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
set\-reos
T}:T{
Enables the termination of reads on reception of the
	end\-of\-string byte for descriptors returned by ibfind()\&.
	Same as setting the REOS bit in a
	ibeos()
	call\&. Default is set\-reos = no\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
set\-xeos
T}:T{
Enables the assertion of EOI on transmission of the
	end\-of\-string byte for descriptors returned by ibfind()\&.
	Same as setting the XEOS bit in a
	ibeos()
	call\&. Default is set\-xeos = no\&.
T}:T{
interface or device
T}:T{
optional
T}
T{
sysfs_device_path
T}:T{
A string which may be used to select a particular piece of hardware by its sysfs device path\&.
T}:T{
interface
T}:T{
optional
T}
T{
timeout
T}:T{
Sets the io timeout for a board or device descriptor
	opened through ibfind()\&.  The possible settings are the same
	as the constants used
	by ibtmo()\&.
	Default is timeout = T3s , which is 3 seconds\&.
T}:T{
interface or device
T}:T{
optional
T}
.TE
.sp 1
.SH "COPYRIGHT"
.br
Copyright \(co 2003-2006, 2008 Frank Mori Hess
.br


ChangeLog
2018-12-25 09:50  dpenkler

	* [r1806] Fix board type

2018-12-25 09:46  dpenkler

	* [r1805] Fix ni_usb_gpib cut & paste error

2018-12-25 09:42  dpenkler

	* [r1804] Simplify usb udev rules as per Frank's suggestion

2018-12-21 10:44  dpenkler

	* [r1803] Changes for multiple board configuration

	  Update INSTALL
	  Add serial to device environment
	  Print serial in gpib_udev_config
	  Add example rules for agilent

2018-12-15 08:55  fmhess

	* [r1802] Add some examples for configuring multiple boards with
	  different minor numbers.

2018-12-15 08:29  fmhess

	* [r1801] Automatically gibe the correct board type to gpib_config
	  in udev rules

2018-12-15 08:01  fmhess

	* [r1800] rename gpib generic rules so it runs at 98
	  (before the driver-specific fules which run at 99)

2018-12-15 07:52  fmhess

	* [r1799] Add fix for mylog from gpib_udev_config to
	  gpib_udev_fxloader.in. Also add some quoting to
	  ensure whitespace is preserved in log messages.

2018-12-14 16:23  dpenkler

	* [r1798] Fix logging in gpib_udev_config for gpib_config options
	  --

2018-12-14 08:30  fmhess

	* [r1797] Removed some bash-isms so scripts work with posix
	  /bin/sh.

2018-12-14 08:11  fmhess

	* [r1796] Revert commit 1793

2018-12-14 08:07  fmhess

	* [r1795] Add --with-udev-libdir option to configure, that I failed
	  to include in earlier commit.

2018-12-13 17:05  dpenkler

	* [r1794] Really remove unused files

2018-12-13 17:02  dpenkler

	* [r1793] Add UDEV_LIBDIR to Makefile.ams\nRemove unused files

2018-12-13 13:44  fmhess

	* [r1792] Added --with-udev-libdir configure option, defaulted
	  to PREFIX/lib/udev.

2018-12-13 12:24  dpenkler

	* [r1791] Rename and move udevadm wrapper to libdir

2018-12-13 12:17  fmhess

	* [r1790] Add missing define of UDEV_LIB_DIR to prevent files from
	  getting installed into root directory.

2018-12-13 12:13  fmhess

	* [r1789] Replaced udev scripts ni_usb_gpib and agilent_82357a with
	  a single generic script.

2018-12-13 11:53  fmhess

	* [r1788] Really delete usermap files this time.

2018-12-13 11:47  fmhess

	* [r1787] Pull out firmware loading into a single udev script
	  gpib_udev_fxloader.

2018-12-13 11:44  dpenkler

	* [r1786] Use #!/bin/sh instead of #!/bin/bash

2018-12-13 11:11  dpenkler

	* [r1785] Add wrapper script for udevadm

2018-12-12 18:04  dpenkler

	* [r1784] Fix udev rules scripts for agilent usb

2018-12-12 15:43  fmhess

	* [r1783] Generate usb script/rules during "make all" rather
	  than possibly waiting until "make install", where the
	  user may be running as root.

2018-12-12 15:27  fmhess

	* [r1782] Fixed assignment that was supposed to be a comparison

2018-12-12 14:03  fmhess

	* [r1781] Removed obsolete hotplug usermaps.

	  Fixed problems with usb udev rules not working if the driver
	  module was not already loaded when the hardware was plugged
	  in.

	  Made usb drivers respond to the sysfs device path of their
	  usb interface rather than the parent usb device. This is helpful
	  since the driver binds to the interfaces rather than the device.

2018-12-12 06:25  fmhess

	* [r1780] Update comments in agilent_82357a udev script to use
	  new --serial and --sysfs-device-path gpib_config options.

2018-12-10 10:49  fmhess

	* [r1779] Added linux/atomic.h compatibility header.

2018-12-10 10:36  fmhess

	* [r1778] Added a couple more devices to ni_usb_gpib.usermap.

2018-12-10 10:28  fmhess

	* [r1777] Add kernel driver support for selecting hardware by sysfs
	  device path
	  or serial number.

2018-12-10 10:27  fmhess

	* [r1776] Add support for selecting hardware by sysfs device path
	  or serial number. Updated gpib_config docs with
	  missing info on some options. Noted in docs that
	  gpib.conf will be in the sysconfdir, which isn't
	  always /etc/ depending on configuration.

2018-12-10 10:11  fmhess

	* [r1775] Add device ids for various ni adapters to
	  ni_usb_gpib.usermap

2018-12-07 09:12  dpenkler

	* [r1774] Remove dependency on autoconf tools
	  Modify INSTALL instructions
	  Update drivers/gpib/Makefile
	  Add Makefile to invoke kernel build scripts
	  Add version.h to drivers/gpib/include
	  Make definitions of smp_mb__xxx_atomic independent of kernel
	  version
	  Add VERBOSE flag for make, default is VERBOSE=0
	  so that build log is more overseeable.
	  Add HAVE_DEV_OF_NODE workaround for CentOS/RHEL7
	  Fix compile warnings in lpvo

2018-12-06 06:10  fmhess

	* [r1773] Reverted r1766 since it was actually comparing the serial
	  number to the device tree path. Linux doesn't seem to
	  have a universal (for all bus types) way of specifying
	  a piece of hardware by path, so we can't use the
	  device tree path to specify a usb device anyways.

2018-12-05 11:19  dpenkler

	* [r1772] Apply gpib_lines patch from Tomislav Ivek

2018-11-29 08:29  dpenkler

	* [r1771] Change barrier() to mb() to be generic

2018-11-28 19:02  dpenkler

	* [r1770] Add kernel version ifdef for smp atomics

2018-11-28 18:02  dpenkler

	* [r1769] Add compat fixes for kernels < 3.15 and 3.16

2018-11-28 14:36  dpenkler

	* [r1768] Add patch from Tomislav Ivek

	  The current linux-gpib Python class can run into a double-close
	  situation. This bug is related to a design point in Gpib.Gpib
	  which
	  cleans up its GPIB handle only on Python's GC cycle. This is not
	  guaranteed to run when the programmer expects it to and in
	  extreme
	  cases might even leak resources. This is resolved by adding a
	  Gpib.close() method for user-controlled cleanup.

2018-11-28 10:26  fmhess

	* [r1767] Applied patch from Derek Kozel <derek.kozel@gmail.com>:

	  During the linux-gpib-kernel make install process the kernel
	  modules_install command is run bit the DESTDIR argument is not
	  passed
	  in meaning that it will always try to install to the system
	  directory
	  rather than a custom directory. Adding
	  INSTALL_MOD_PATH="$(DESTDIR)" to
	  the command correctly passes the information.

2018-11-16 20:30  fmhess


linux-gpib-user/language/guile/gpib.c
/*
 * gpib.c - guile binding for LinuxGpib
 *
 * Copyright (C) 2003 Stefan Jahn <stefan@lkcc.org>
 *
 * LinuxGpib is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * LinuxGpib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with LinuxGpib; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <guile/gh.h>
#include <gpib/ib.h>

#include "gpib.h"

#define FUNC_NAME "ibdev"
SCM
guile_ibdev (SCM board_index, SCM pad, SCM sad, SCM timeout, SCM eoi, SCM eos)
{
  int fd, _eoi;

  SCM_ASSERT_TYPE (SCM_EXACTP (board_index), board_index,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (pad), pad,
		   SCM_ARG2, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (sad), sad,
		   SCM_ARG3, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (timeout), timeout,
		   SCM_ARG4, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (eoi) || SCM_BOOLP (eoi), eoi,
		   SCM_ARG5, FUNC_NAME, "exact or bool");
  SCM_ASSERT_TYPE (SCM_EXACTP (eos), eos,
		   SCM_ARG6, FUNC_NAME, "exact");

  if (SCM_BOOLP (eoi))
    _eoi = SCM_NFALSEP (eoi);
  else
    _eoi = SCM_NUM2INT (SCM_ARG5, eoi);


  if ((fd = ibdev (SCM_NUM2INT (SCM_ARG1, board_index),
		   SCM_NUM2INT (SCM_ARG2, pad),
		   SCM_NUM2INT (SCM_ARG3, sad),
		   SCM_NUM2INT (SCM_ARG4, timeout),
		   _eoi,
		   SCM_NUM2INT (SCM_ARG6, eos))) < 0) {
    scm_syserror (FUNC_NAME);
  }
  return scm_int2num (fd);
}
#undef FUNC_NAME

#define FUNC_NAME "ibwrt"
SCM
guile_ibwrt (SCM ud, SCM data)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_STRINGP (data), data,
		   SCM_ARG2, FUNC_NAME, "string");

  if ((ret = ibwrt (SCM_NUM2INT (SCM_ARG1, ud),
		    SCM_STRING_CHARS (data),
		    SCM_NUM2INT (SCM_ARG2,
				 scm_string_length (data)))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibcmd"
SCM
guile_ibcmd (SCM ud, SCM commands)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_STRINGP (commands), commands,
		   SCM_ARG2, FUNC_NAME, "string");

  if ((ret = ibcmd (SCM_NUM2INT (SCM_ARG1, ud),
		    SCM_STRING_CHARS (commands),
		    SCM_NUM2INT (SCM_ARG2,
				 scm_string_length (commands)))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibrd"
SCM
guile_ibrd (SCM ud, SCM bytes)
{
  int ret, len;
  char *data;
  SCM val;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (bytes), bytes,
		   SCM_ARG2, FUNC_NAME, "exact");
  len = SCM_NUM2INT (SCM_ARG2, bytes);

  if ((data = (char *) malloc (len)) == NULL) {
    scm_memory_error (FUNC_NAME);
  }

  if ((ret = ibrd (SCM_NUM2INT (SCM_ARG1, ud), data, len)) & ERR) {
    free (data);
    return SCM_BOOL_F;
  }

  val = scm_mem2string (data, ibcnt);
  free (data);
  return val;
}
#undef FUNC_NAME

#define FUNC_NAME "ibfind"
SCM
guile_ibfind (SCM name)
{
  char *str;
  int ret;

  SCM_ASSERT_TYPE (SCM_STRINGP (name), name,
		   SCM_ARG1, FUNC_NAME, "string");
  str = guile_to_string (name);

  if ((ret = ibfind (str)) < 0) {
    free (str);
    return SCM_BOOL_F;
  }
  free (str);
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibsre"
SCM
guile_ibsre (SCM ud, SCM enable)
{
  int ret, val;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (enable) || SCM_BOOLP (enable), enable,
		   SCM_ARG2, FUNC_NAME, "exact or bool");

  if (SCM_BOOLP (enable))
    val = SCM_NFALSEP (enable);
  else
    val = SCM_NUM2INT (SCM_ARG2, enable);

  if ((ret = ibsre (SCM_NUM2INT (SCM_ARG1, ud), val)) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibsic"
SCM
guile_ibsic (SCM ud)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");

  if ((ret = ibsic (SCM_NUM2INT (SCM_ARG1, ud))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibclr"
SCM
guile_ibclr (SCM ud)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");

  if ((ret = ibclr (SCM_NUM2INT (SCM_ARG1, ud))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibonl"
SCM
guile_ibonl (SCM ud, SCM online)
{
  int ret, val;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (online) || SCM_BOOLP (online), online,
		   SCM_ARG2, FUNC_NAME, "exact or bool");

  if (SCM_BOOLP (online))
    val = SCM_NFALSEP (online);
  else
    val = SCM_NUM2INT (SCM_ARG2, online);

  if ((ret = ibonl (SCM_NUM2INT (SCM_ARG1, ud), val)) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibwait"
SCM
guile_ibwait (SCM ud, SCM status_mask)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (status_mask), status_mask,
		   SCM_ARG2, FUNC_NAME, "exact");

  if ((ret = ibwait (SCM_NUM2INT (SCM_ARG1, ud),
		     SCM_NUM2INT (SCM_ARG2, status_mask))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibrsp"
SCM
guile_ibrsp (SCM ud)
{
  int ret;
  char poll[16];

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");

  if ((ret = ibrsp (SCM_NUM2INT (SCM_ARG1, ud), poll)) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (poll[0]);
}
#undef FUNC_NAME

#define FUNC_NAME "ibtrg"
SCM
guile_ibtrg (SCM ud)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");

  if ((ret = ibtrg (SCM_NUM2INT (SCM_ARG1, ud))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "ibrsv"
SCM
guile_ibrsv (SCM ud, SCM status_byte)
{
  int ret;

  SCM_ASSERT_TYPE (SCM_EXACTP (ud), ud,
		   SCM_ARG1, FUNC_NAME, "exact");
  SCM_ASSERT_TYPE (SCM_EXACTP (status_byte), status_byte,
		   SCM_ARG2, FUNC_NAME, "exact");

  if ((ret = ibrsv (SCM_NUM2INT (SCM_ARG1, ud),
		    SCM_NUM2INT (SCM_ARG2, status_byte))) & ERR) {
    return SCM_BOOL_F;
  }
  return scm_int2num (ret);
}
#undef FUNC_NAME

#define FUNC_NAME "iberr"
SCM guile_iberr (void)
{
  return scm_int2num (iberr);
}
#undef FUNC_NAME

#define FUNC_NAME "ibcnt"
SCM guile_ibcnt (void)
{
  return scm_long2num (ibcntl);
}
#undef FUNC_NAME
#pragma GCC diagnostic warning "-Wincompatible-pointer-types"
void
gpib_bindings (void)
{
  scm_c_define_gsubr ("ibdev",  6, 0, 0, guile_ibdev);
  scm_c_define_gsubr ("ibwrt",  2, 0, 0, guile_ibwrt);
  scm_c_define_gsubr ("ibcmd",  2, 0, 0, guile_ibcmd);
  scm_c_define_gsubr ("ibrd",   2, 0, 0, guile_ibrd);
  scm_c_define_gsubr ("ibfind", 1, 0, 0, guile_ibfind);
  scm_c_define_gsubr ("ibsre",  2, 0, 0, guile_ibsre);
  scm_c_define_gsubr ("ibsic",  1, 0, 0, guile_ibsic);
  scm_c_define_gsubr ("ibclr",  1, 0, 0, guile_ibclr);
  scm_c_define_gsubr ("ibonl",  2, 0, 0, guile_ibonl);
  scm_c_define_gsubr ("ibwait", 2, 0, 0, guile_ibwait);
  scm_c_define_gsubr ("ibrsp",  1, 0, 0, guile_ibrsp);
  scm_c_define_gsubr ("ibtrg",  1, 0, 0, guile_ibtrg);
  scm_c_define_gsubr ("ibrsv",  1, 0, 0, guile_ibrsv);
  scm_c_define_gsubr ("iberr",  0, 0, 0, guile_iberr);
  scm_c_define_gsubr ("ibcnt",  0, 0, 0, guile_ibcnt);

  scm_c_define ("DCAS",  scm_int2num (DCAS));
  scm_c_define ("DTAS",  scm_int2num (DTAS));
  scm_c_define ("LACS",  scm_int2num (LACS));
  scm_c_define ("TACS",  scm_int2num (TACS));
  scm_c_define ("ATN",   scm_int2num (ATN));
  scm_c_define ("CIC",   scm_int2num (CIC));
  scm_c_define ("REM",   scm_int2num (REM));
  scm_c_define ("LOK",   scm_int2num (LOK));
  scm_c_define ("CMPL",  scm_int2num (CMPL));
  scm_c_define ("EVENT", scm_int2num (EVENT));
  scm_c_define ("SPOLL", scm_int2num (SPOLL));
  scm_c_define ("RQS",   scm_int2num (RQS));
  scm_c_define ("SRQI",  scm_int2num (SRQI));
  scm_c_define ("END",   scm_int2num (END));
  scm_c_define ("TIMO",  scm_int2num (TIMO));
  scm_c_define ("ERR",   scm_int2num (ERR));

  scm_c_define ("GTL", scm_int2num (GTL));
  scm_c_define ("SDC", scm_int2num (SDC));
  scm_c_define ("PPC", scm_int2num (PPC));
  scm_c_define ("GET", scm_int2num (GET));
  scm_c_define ("TCT", scm_int2num (TCT));
  scm_c_define ("LLO", scm_int2num (LLO));
  scm_c_define ("DCL", scm_int2num (DCL));
  scm_c_define ("PPU", scm_int2num (PPU));
  scm_c_define ("SPE", scm_int2num (SPE));
  scm_c_define ("SPD", scm_int2num (SPD));
  scm_c_define ("UNL", scm_int2num (UNL));
  scm_c_define ("UNT", scm_int2num (UNT));
  scm_c_define ("PPD", scm_int2num (PPD));

  scm_c_define ("EDVR", scm_int2num (EDVR));
  scm_c_define ("ECIC", scm_int2num (ECIC));
  scm_c_define ("ENOL", scm_int2num (ENOL));
  scm_c_define ("EADR", scm_int2num (EADR));
  scm_c_define ("EARG", scm_int2num (EARG));
  scm_c_define ("ESAC", scm_int2num (ESAC));
  scm_c_define ("EABO", scm_int2num (EABO));
  scm_c_define ("ENEB", scm_int2num (ENEB));
  scm_c_define ("EOIP", scm_int2num (EOIP));
  scm_c_define ("ECAP", scm_int2num (ECAP));
  scm_c_define ("EFSO", scm_int2num (EFSO));
  scm_c_define ("EBUS", scm_int2num (EBUS));
  scm_c_define ("ESTB", scm_int2num (ESTB));
  scm_c_define ("ESRQ", scm_int2num (ESRQ));
  scm_c_define ("ETAB", scm_int2num (ETAB));

  scm_c_define ("REOS", scm_int2num (REOS));
  scm_c_define ("XEOS", scm_int2num (XEOS));
  scm_c_define ("BIN",  scm_int2num (BIN));

  scm_c_define ("TNONE",  scm_int2num (TNONE));
  scm_c_define ("T10us",  scm_int2num (T10us));
  scm_c_define ("T30us",  scm_int2num (T30us));
  scm_c_define ("T100us", scm_int2num (T100us));
  scm_c_define ("T300us", scm_int2num (T300us));
  scm_c_define ("T1ms",   scm_int2num (T1ms));
  scm_c_define ("T3ms",   scm_int2num (T3ms));
  scm_c_define ("T10ms",  scm_int2num (T10ms));
  scm_c_define ("T30ms",  scm_int2num (T30ms));
  scm_c_define ("T100ms", scm_int2num (T100ms));
  scm_c_define ("T300ms", scm_int2num (T300ms));
  scm_c_define ("T1s",    scm_int2num (T1s));
  scm_c_define ("T3s",    scm_int2num (T3s));
  scm_c_define ("T10s",   scm_int2num (T10s));
  scm_c_define ("T30s",   scm_int2num (T30s));
  scm_c_define ("T100s",  scm_int2num (T100s));
  scm_c_define ("T300s",  scm_int2num (T300s));
  scm_c_define ("T1000s", scm_int2num (T1000s));
}

extern void
gpib_init ()
{
  scm_add_feature ("gpib");
  gpib_bindings ();
}
